fix(SFTP): infinite loop scanning directories with upward symlinks

leftybournes opened the pull request:
<!--
  - 🚨 SECURITY INFO
  -
  - Before sending a pull request that fixes a security issue please report it via our HackerOne page (https://hackerone.com/nextcloud) following our security policy (https://nextcloud.com/security/). This allows us to coordinate the fix and release without potentially exposing all Nextcloud servers and users in the meantime.
-->

* Resolves: #52214 <!-- related github issue -->

## Summary
When a symlink is detected, check if the path already includes the file. If it does, ignore the file. This helps prevents the scanner from going into an infinite loop when the configured root contains a directory symlinked to a parent/grandparent/higher directory which eventually contains it.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [X] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

Commit:
fix(SFTP): infinite loop scanning directories with upward symlinks

Signed-off-by: Kent Delante <kent.delante@proton.me>

provokateurin commented on apps/files_external/lib/Lib/Storage/SFTP.php:
```diff
@@ -252,15 +252,23 @@ public function rmdir(string $path): bool {
+				$fileAdded = false;
+
+				if (!empty($linkedPath)) {
+					$fileAdded = substr_count($absPath, basename($absPath) . '/' . $file) > 0;
```
I think using `substr_count` is quite unsafe here.
It does not check that the linked path really is part of the parent path.
It can also lead to other funny problems when there are files and folders with similar names and your check would confuse a symlink at `/dir123/sym` to a file at `/dir` to cause the loop with `/dir123` while it actually doesn't.

Besides that I'm not sure what you are checking here, since you are not involving the `$linkedPath` and it seems to me like this will always evaluate to true (but I might be wrong as I don't know how this exactly works).

leftybournes replied:
I'm comparing the symlink along with each child (e.g. `sym/child_dir`) done in `basename($absPath) . '/' . $file` against the current absolute path and it only checks if the current path is a symlink. Would the issue persist in that case?

From my testing, `$linkedPath` is a blank string when it's checked against a regular file or directory so checking if it's not empty evaluates to false in those cases as expected.

I see that my code lacks clarity. I'll make adjustments accordingly.

guerby replied:
Hi, thanks for the patch. 

Thinking a bit more about it I think it's best to just ignore symlinks altogether. Following symlinks will give multiple path to get to the same file which will be confusing (and filling the database). Also I don't think a purely local algorithm can in all case avoid loops, eg you can have a symlink /dir/d1/a to /dir/d2 and inside /dir/d2 have a symlink b to /dir/d1.

Ignoring symlinks will probably index all files anyway in the external storage as behind a symlink lies a real directory.

What do you think?

provokateurin reviewed (changes requested):
Some tests for this would be great :)

leftybournes commented:
> Some tests for this would be great :)

Sure, I'll add tests for this.

Commit:
test(SFTP): Setup authentication and symlink tests

Signed-off-by: Kent Delante <kent.delante@proton.me>

Commit:
fix(SFTP): Check each symlink directory if it repeats in path

The absolute path currently always starts with a '/' so check the
current directory along with each symlinked child if it is already in
the path (e.g. /parent/symlink/). Adding '/' to both ends helps with
avoiding collisions with partial matches in the path.

Signed-off-by: Kent Delante <kent.delante@proton.me>

Commit:
chore: Fix SFTP and SftpTest formatting with php-cs-fixer

Signed-off-by: Kent Delante <kent.delante@proton.me>

Commit:
test(SFTP): Use common storage abstractions to verify implementation

Signed-off-by: Kent Delante <kent.delante@proton.me>

Commit:
chore: Fix sftp test formatting

Signed-off-by: Kent Delante <kent.delante@proton.me>

provokateurin commented on apps/files_external/tests/Storage/SftpTest.php:
```diff
@@ -133,4 +134,37 @@ public function configProvider() {
 		];
 	}
+
+	public function testAuthenticated(): void {
+		$this->assertTrue($this->instance->getConnection()->isAuthenticated());
+	}
```
Should not be necessary, but might be nice to add to `setUp` so any test fails early and not with a confusing error message that is not related to the actual test.

leftybournes replied:
It makes sense to just remove it anyway since `setUp` already has a `mkdir()` call which requires the client to be authenticated.

provokateurin commented on apps/files_external/lib/Lib/Storage/SFTP.php:
```diff
@@ -252,15 +252,19 @@ public function rmdir(string $path): bool {
-				if ($file !== '.' && $file !== '..') {
+			foreach ($list as $file => $stat) {
+				$fileAdded = ((int)$stat['type'] === NET_SFTP_TYPE_SYMLINK) &&
+							 str_contains($absPath, '/' . basename($absPath) . '/' . $file . '/');
```
Can you explain the comparison? I'm still not sure what exactly it tries to test and how.

leftybournes replied:
It takes the current directory and the symlink and checks if they're in the current absolute path. Assuming that the current absolute path is `/home/user/data` and it contains a symlink `slink`, which points to `data`. It checks if the string `'/data/slink/'` is in `'/home/user/data'`. Just trying to prevent it from repeating itself and turning into something like `/home/user/data/slink/data/slink/` or something similar.

I realize it might still be a wonky solution. I did some further testing and `readlink()` gives me the absolute path the symlink points to (I didn't think it did previously and I realize it's because I didn't use it properly). I'll probably use that and compare it to the absolute path instead. Any thoughts?

provokateurin reviewed (commented):
The test looks good, but I need some clarification on the check you are doing.
